Add JWT support

sapkra opened the pull request:
With this PR I'm adding support for JWT tokens to the prosody container.
I don't wanted to copy the prosody modules out of the `jitsi/jitsi-meet` repository so that I added it as a submodule to have no duplicates and one single point of truth.

I hope I didn't missed something.

@amirasaran Thanks for your help.

sapkra commented on prosody/Dockerfile:
```diff
@@ -2,11 +2,19 @@ FROM jitsi/base
+	luarocks install luaossl  CRYPTO_LIBDIR=/usr/lib/x86_64-linux-gnu  CRYPTO_INCDIR=/usr/include OPENSSL_LIBDIR=/usr/lib/x86_64-linux-gnu OPENSSL_INCDIR=/usr/include && \
+	luarocks install luajwt CRYPTO_LIBDIR=/usr/lib/x86_64-linux-gnu  CRYPTO_INCDIR=/usr/include OPENSSL_LIBDIR=/usr/lib/x86_64-linux-gnu OPENSSL_INCDIR=/usr/include && \
+	luarocks install luajwtjitsi CRYPTO_LIBDIR=/usr/lib/x86_64-linux-gnu  CRYPTO_INCDIR=/usr/include OPENSSL_LIBDIR=/usr/lib/x86_64-linux-gnu OPENSSL_INCDIR=/usr/include && \
+	luarocks install jwt-jitsi CRYPTO_LIBDIR=/usr/lib/x86_64-linux-gnu  CRYPTO_INCDIR=/usr/include OPENSSL_LIBDIR=/usr/lib/x86_64-linux-gnu OPENSSL_INCDIR=/usr/include
```
Do I really need to add `CRYPTO_LIBDIR`, `CRYPTO_INCDIR`, `OPENSSL_LIBDIR` and `OPENSSL_INCDIR `?
I saw it in the issue and just copied it.

saghul replied:
Please try without them and simplify the code if they are not needed.

aaronkvanmeerten replied:
the OPENSSL and CRYPTO dirs are likely required when building the jwt code, as components of these libraries are used when verifying signatures on JWT tokens.

saghul replied:
@aaronkvanmeerten sure, but being in standard locations, the env variables may be not required?

sapkra replied:
Updated the PR and removed these variables - I tested it without them and it's working fine.

sapkra commented on prosody/Dockerfile:
```diff
@@ -2,11 +2,19 @@ FROM jitsi/base
+COPY git-modules/jitsi-meet/resources/prosody-plugins /prosody-plugins
+
+RUN \
+	luarocks list --outdated && \
```
I think I can remove this line because there are always no rocks installed at this time. Right?
Sorry I'm new in lua.

saghul replied:
I don't think it's needed in this setup.

ronanmcnulty replied:
git-modules/jitsi-meet/resources/prosody-plugins is empty, the copy fails

saghul commented on prosody/Dockerfile:
```diff
@@ -2,11 +2,19 @@ FROM jitsi/base
 RUN \
 	apt-dpkg-wrap apt-get update && \
-	apt-dpkg-wrap apt-get install -t stretch-backports -y prosody && \
+	apt-dpkg-wrap apt-get install -t stretch-backports -y prosody luarocks gcc libssl1.1 libssl1.0-dev git && \
```
do we want to install these new packages from backports too? why?

sapkra replied:
Yeah I think we don't need to do that. I will fix it in my next update.

Rodrigora commented:
FYI I just tested this PR and worked fine 👍

damencho commented:
There is a patch about an event for bosh-session. Without it jwt will not work. And that change is available only in trunk.
We are using for meet.jit.si trunk version 747 for a long time.

sapkra commented:
Thanks for this information. I tried running prosody-trunk `1018` but I always got the error `token required`. Now I installed `747` and it's working fine.

sapkra commented on prosody/rootfs/defaults/conf.d/jitsi-meet.cfg.lua:
```diff
@@ -37,9 +53,15 @@ Component "{{ .Env.XMPP_INTERNAL_MUC_DOMAIN }}" "muc"
     storage = "none"
+    {{ if .Env.ENABLE_TOKEN_AUTH }}
+    modules_enabled = { "token_verification" }
+    {{ end }}
```
One more question: Is it right to add this module on both `XMPP_INTERNAL_MUC_DOMAIN ` and `XMPP_MUC_DOMAIN `?

saghul replied:
It should only be needed in XMPP_MUC_DOMAIN

saghul commented on prosody/Dockerfile:
```diff
@@ -1,12 +1,24 @@
+	luarocks install luabitop
 
 COPY rootfs/ /
+COPY git-modules/jitsi-meet/resources/prosody-plugins /prosody-plugins
```
Using a git submodule just for this is quite overkill. Please don't. use ADD for the plugins we specifically want instead. (we don't need all the plugins there)

sapkra replied:
Do you know which plugins we need? I'm sure that we need `token/util.lib.lua` and `mod_token_verification.lua`. Do we need more e.g. `mod_token_auth.lua`?

saghul replied:
I believe we do, yes.

sapkra replied:
To apply this change I need a full list of all necessary plugins we need.

saghul replied:
mod_auth_token
mod_token_verification
token/

sapkra replied:
Updated the PR - please have a look.

saghul commented:
> There is a patch about an event for bosh-session. Without it jwt will not work. And that change is available only in trunk.
> We are using for meet.jit.si trunk version 747 for a long time.

@damencho The code in our patch (https://github.com/jitsi/jitsi-meet/blob/master/resources/prosody-plugins/mod_bosh.lua.patch) is already part of the 0.10.2 release, which is what we are using, because we are pulling it from stretch-backports: https://packages.debian.org/stretch-backports/prosody

So, do we really need to use prosody-trunk here?

@sapkra Did you test if this war working before switching to prosody-trunk? I'm afraid the docs may need an update.

damencho commented:
@saghul if we are talking just for jwt maybe it is ok. There are some muc events that were only available in trunk.

saghul commented:
@damencho What MUC events? What do we use them for? I'd like to take the opportunity and document this properly.

damencho commented:
So for tokens I see just `module:hook("muc-room-pre-create"` and `module:hook("muc-occupant-pre-join"`. For other modules we use more of those.

saghul commented:
@damencho Thanks! Ok, it seems we need Prosody 0.11.X at least then. AFAIS Prosody does provide a Debian repo: https://prosody.im/download/package_repository so that should be the best way forward IMHO.

damencho commented:
I have in my todo list to test 11, cause somebody reported not working (he was testing jwt).

saghul commented:
Sounds like this PR is a good opportunity then! @sapkra can you please update so we can test this? Cheers!

sapkra commented:
I was the guy who reported that it's not working with prosody 0.11.0. :smile:
I updated the Dockerfile already locally. I will push these changes in the next hours.

saghul commented:
@sapkra Haha we came full-circle then :-) No rush, let's get this done The Right Way (TM).

sapkra commented:
I tested it again with 0.11.1 & I'm still getting the error I mentioned in https://github.com/jitsi/lib-jitsi-meet/issues/837

saghul commented:
@sapkra Thanks! Once this PR is complete we'll have an easy way to test / debug this.

sapkra commented:
I think the PR is complete now - feel free to test/debug it.

saghul commented on prosody/Dockerfile:
```diff
@@ -1,10 +1,25 @@
 FROM jitsi/base
 
+ADD https://packages.prosody.im/debian/pool/main/p/prosody/prosody_0.11.1-1~stretch1_amd64.deb /tmp/prosody.deb
```
Please add the repo instead of pulling the file manually, this way we'll get automatic updates as they are pushed to the repo.

sapkra replied:
Ohh yeah you're right. I changed it.

saghul commented on prosody/Dockerfile:
```diff
@@ -1,10 +1,25 @@
 RUN \
 	apt-dpkg-wrap apt-get update && \
-	apt-dpkg-wrap apt-get install -t stretch-backports -y prosody && \
+	apt-dpkg-wrap apt-get install -y lua5.2 liblua5.2-dev lua-expat lua-socket lua-filesystem ssl-cert luarocks gcc libssl1.1 libssl1.0-dev git && \
```
Why is git necessary?

sapkra replied:
For me it seems like git is used to download `luajwt`.

````
Installing https://luarocks.org/luajwtjitsi-1.3-7.rockspec
Cloning into 'luajwt'...
Note: checking out 'e7a6ca8cbe5941638db34e280eff7e75c07c4ec9'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by performing another checkout.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -b with the checkout command again. Example:

  git checkout -b <new-branch-name>
````

saghul replied:
👍

saghul commented on prosody/Dockerfile:
```diff
@@ -1,10 +1,30 @@
 RUN \
 	apt-dpkg-wrap apt-get update && \
-	apt-dpkg-wrap apt-get install -t stretch-backports -y prosody && \
+	apt-dpkg-wrap apt-get install -y gnupg lua5.2 liblua5.2-dev lua-expat lua-socket lua-filesystem ssl-cert luarocks gcc libssl1.1 libssl1.0-dev git
```
Please remove gnupg  (and I think all -dev libraries too) to keep the container smaller.

sapkra replied:
Gnupg is needed to add the key and there was another problem with the upgrade to lua 5.2 so that I had to add the dev package.
I can try it without them again.
Gnupg can be removed after adding the key.

saghul replied:
Sorry, I meant remove them at the end, after the installation is done :-) They are probably needed at build time indeed.

sapkra replied:
Alright. It's pushed.

saghul commented on prosody/Dockerfile:
```diff
@@ -1,10 +1,30 @@
+	luarocks install luajwt && \
+	luarocks install luajwtjitsi && \
+	luarocks install jwt-jitsi && \
+	luarocks install luasec && \
```
Let's use the debian package: https://packages.debian.org/source/stretch/lua-sec

saghul commented on prosody/Dockerfile:
```diff
@@ -1,10 +1,30 @@
+	luarocks install luajwtjitsi && \
+	luarocks install jwt-jitsi && \
+	luarocks install luasec && \
+	luarocks install lua-cjson 2.1.0-1
```
Let's use the debian package: https://packages.debian.org/stretch/lua-cjson

sapkra replied:
But because we are using lua 5.2 now I have to lock the version to 2.1.0-1. Otherwise I get this error while building.

```
lua_cjson.c: In function 'json_append_data':
lua_cjson.c:743:19: warning: implicit declaration of function 'lua_objlen' [-Wimplicit-function-declaration]
             len = lua_objlen(l, -1);
```

And this one while running the container:
```
error loading module 'cjson' from file '/usr/local/lib/lua/5.2/cjson.so':
/usr/local/lib/lua/5.2/cjson.so: undefined symbol: lua_objlen
```

saghul replied:
Do you see this error when using the Debian packages themselves?

sapkra replied:
Yes, I think I copied these errors from exactly this build.

saghul replied:
Looks like you still have some old module there. Debian won't install the package in /usr/local

sapkra replied:
I just noticed that luajwt is also installing it as a dependency. Maybe the log was from this part of the build process.

saghul replied:
Hum, that's odd, why on earth doesn't it detect it's installed? Maybe because it's not installed via luarocks?

saghul commented on prosody/Dockerfile:
```diff
@@ -1,10 +1,30 @@
+	rm -rf /etc/prosody /tmp/key
+
+RUN \
+	luarocks install luaossl && \
```
Let's use the Debian package: https://packages.debian.org/stretch/lua-luaossl

saghul commented on prosody/Dockerfile:
```diff
@@ -1,10 +1,30 @@
 
+ADD https://prosody.im/files/prosody-debian-packages.key /tmp/key
+
+ADD https://raw.githubusercontent.com/jitsi/jitsi-meet/master/resources/prosody-plugins/token/util.lib.lua /prosody-plugins/token/util.lib.lua
```
Please pin this to a specific commit, so we can control what version of the file is pulled.

sapkra replied:
Done.

saghul commented:
We have been quite busy this week, sorry. Didn’t have any time at all to look at this.

sapkra commented:
Are you still busy or is it possible to find some time?
It would be great to get this tested & merged soon - it will simplify a lot in my project.

sapkra commented on docker-compose.yml:
```diff
@@ -15,6 +15,11 @@ services:
+            - JWT_APP_ID
+            - JWT_APP_SECRET
+            - JWT_ACCEPTED_ISSUERS
+            - JWT_ACCEPTED_AUDIENCES
```
I noticed that I set the variables for the wrong container. They has to be added to the prosody container.

saghul commented:
>     1. Prosody 0.11.x is saving the certs in the `data_path` so I have to edit the script. I would prefer to let them at this path but it would result in a more or less breaking change. What's your opinion?

I'd set that to /config/keys

As for breaking, this repo is contains "experimental" in all caps, so it's fair game :-)



>     1. I'm getting the following error. To fix this the `storage` should be changed to `internal` where we have set the `muc_room_cache_size` in `jitsi-meet.cfg.lua`. [jitsi/jitsi-meet#3454](https://github.com/jitsi/jitsi-meet/issues/3454)

The thing is we want to have no storage at all. Maybe it has a different name in 0.11 ?

Thanks for continuing your work on this!

saghul commented:
@sapkra Hey Paul, after all your latest changes, are you able to get it working on 0.11 ?

sapkra commented:
No, I just applied the changes I already had in my test environment. I also tested it again and I got the same error.

saghul commented:
@sapkra Thanks for confirming.

saghul commented:
@sapkra Thanks! Some things must have become simpler, since Prosody is now already at 0.11.

sapkra commented:
Ohh, I didn't noticed that prosody 0.11 is already published in stretch-backports. I will simplify the Dockerfile.

sapkra commented:
I refactored the Dockerfile and also removed `luarocks install jwt-jitsi` from it.
I think we don't need it because it is also not getting installed within the installation process of the `jitsi-meet-tokens` debian package (I couldn't find it).

saghul commented on prosody/Dockerfile:
```diff
@@ -1,11 +1,24 @@
+RUN \
+  luarocks install lua-cjson 2.1.0-1 && \
+	luarocks install luajwtjitsi && \
+	luarocks install basexx
```
This one we have on Debian: https://packages.debian.org/stretch/lua-basexx

sapkra replied:
I wanted to use the same methods like `jitsi-meet-tokens`.
https://github.com/jitsi/jitsi-meet/blob/master/debian/jitsi-meet-tokens.postinst#L68

Why do we want to do it differently?

saghul replied:
If it works as expected I'll also make the change in the jitsi-meet-tokens package.

sapkra replied:
Done.

saghul commented on prosody/Dockerfile:
```diff
@@ -1,11 +1,24 @@
 	rm -rf /etc/prosody
 
+RUN \
+  luarocks install lua-cjson 2.1.0-1 && \
```
I hate this cjson thing so much I took a closer look. It's totally unmaintained and there is no fork which supports Lua 5.2 properly (unless you se an old version which didn't implement certain feature). So it doesn't seem very future proof.

Since the luajwtjitsi module is ours, I think we can just swap the cjson module for another better JSON library.

In addition, luacrypto is super-deprecated and it's recommended to use luaossl now.

I'm going to try to update luajwtjitsi so it uses those 2 so this can be greatly simplified. You don't have to do anything, but thanks for listening to my rant :-)

saghul commented:
Ok, this looks pretty much ready to test and dig why it's not working with the update to 0.11. I'll get on it soon.

FYI, I re-targetted your PR to the `dev` branch. I want to try a new way of doing things here: we merge the stuff to the dev branch and merge dev into master when a new release to DockerHub is made.

saghul commented:
@sapkra Now that we know that the problem was, can you please embed the required JWT files in this PR (in defaults/jwt for example) and we'll remove them later, once we have this in jitsi-meet.

sapkra commented:
I just added an `sed` command to the Dockerfile which will replace `hook` through `hook_global`. I think this is the cleanest temporary solution.

saghul commented:
> I just added an `sed` command to the Dockerfile which will replace `hook` through `hook_global`. I think this is the cleanest temporary solution.

Yeah, sounds good! I hope to merge this really soon. Thanks for your patience!

sapkra commented:
@saghul There are still two open discussions related to `lua-cjson`. Do you want to refactor this stuff after we merged the PR?

saghul commented:
> There are still two open discussions related to `lua-cjson`. Do you want to refactor this stuff after we merged the PR?

Yes. I have no time to address those now, but I think this PR is very valuable as is, so we can address those later.

saghul commented:
@sapkra I sarted testing this and ran into some issues: the prosody plugin paths are owned by root, so prosody cznnot read them. I solved by adding the following:

```
diff --git a/prosody/rootfs/etc/cont-init.d/10-config b/prosody/rootfs/etc/cont-init.d/10-config
index 45efdbb..01b4925 100644
--- a/prosody/rootfs/etc/cont-init.d/10-config
+++ b/prosody/rootfs/etc/cont-init.d/10-config
@@ -10,6 +10,14 @@ if [[ "$(stat -c %U /config)" != "prosody" ]]; then
     chown -R prosody /config
 fi
 
+if [[ "$(stat -c %U /prosody-plugins)" != "prosody" ]]; then
+    chown -R prosody /prosody-plugins
+fi
+
+if [[ "$(stat -c %U /prosody-plugins-custom)" != "prosody" ]]; then
+    chown -R prosody /prosody-plugins-custom
+fi
+
 if [[ ! -f $PROSODY_CFG ]]; then
     cp -r /defaults/* /config
     tpl /defaults/conf.d/jitsi-meet.cfg.lua > /config/conf.d/jitsi-meet.cfg.lua

```

Can you please add this changes to the PR? Thank you!

sapkra commented:
You are also able to push it. `Allow edits from maintainers.` is checked. 😉

saghul commented:
> You are also able to push it. `Allow edits from maintainers.` is checked. 😉

Oh, I forgot to check.



> OpenShift is starting the container always with a user which is in the root group - so I didn't noticed this problem.

No worries. Also, good to know!

saghul commented:
Thanks for all your work here @sapkra, it works beautifully! I'll hold on making a new release until I make some changes I hope to make soon: introduce some AUTH_TYPE wich can be "internal" (the default) "jwt" and soon "ldap".

alexanderankin commented:
if its not too much to ask, what is the use case of this? i know how you can have ldap server, and user accounts on it, but jwt is like crypto signiature?

saghul commented:
This is the typical auth scenario when you are embedding Jitsi into an existing app which has its own authentication. You’d create a token in your backend and configure Jitsi to use JWT for authentication. The token can limit the room that a user can join, establish their saben and avatar on the Jitsi side of things in addition.

alexanderankin commented:
so i have to link to the page with "myinstance.mydomain.com/theroom?token=eyJhbGci..."? where is token expected? in "Authorization: Bearer eyJhbGci..."? I am used to seeing this in API call not in page. Or am I supposed to integrate much closer, like consume video feed stream and show to user in my own app with html5 technologies?

saghul commented:
You can use the URL parameter, yes, or our iframe API, in you are embedding Jitsi into an existing site / application.

valsaraj commented:
Following error when try to install JWT token with latest jitsi-meet:

> apt-get install jitsi-meet-tokens
> Reading package lists... Done
> Building dependency tree       
> Reading state information... Done
> Some packages could not be installed. This may mean that you have
> requested an impossible situation or if you are using the unstable
> distribution that some required packages have not yet been created
> or been moved out of Incoming.
> The following information may help to resolve the situation:
> 
> The following packages have unmet dependencies:
>  jitsi-meet-tokens : Depends: prosody-trunk (>= 1nightly747) but it is not installable or
>                               prosody-0.11 but it is not installable or
>                               prosody (>= 0.11.2) but 0.10.0-1build1 is to be installed

saghul commented:
You don’t need to install that if you are using the Docker setup.

valsaraj commented:
Would you share the link of Docker referred?
Why it's different from the one comes with quick install?

saghul commented:
You are commenting in an issue on the Docker setup repo. Hence we will assume you are using the Docker setup. Please open a discussion topic in our community otherwise.
